### ISSUE TITLE
Pattern matcher does not return all permutations when it should

### ISSUE THREAD

**Issue by ngeiswei** [opencog/atomspace#2346]
Provided the following KB

```scheme
(List (Concept "A") (Set (Predicate "P") (Predicate "Q")))
```

and the following query

```scheme
(define query-ok
  (Bind
    (Present (Set (Variable "$X") (Variable "$Y")))
    (Implication (Variable "$X") (Variable "$Y"))))
```

the pattern matcher, with `(cog-execute! query-ok)`, returns

```scheme
(SetLink
   (ImplicationLink
      (PredicateNode "P")
      (PredicateNode "Q")
   )
   (ImplicationLink
      (PredicateNode "Q")
      (PredicateNode "P")
   )
)
```

as it should.

However, for the following query

```scheme
(define query-not-ok
  (Bind
    (Present (List (Variable "$Z") (Set (Variable "$X") (Variable "$Y"))))
    (Implication (Variable "$X") (Variable "$Y"))))
```

it returns, with `(cog-execute! query-not-ok)`

```scheme
(SetLink
   (ImplicationLink
      (PredicateNode "P")
      (PredicateNode "Q")
   )
)
```

as it should not.

**Comment by ngeiswei** [comment]
Some more weirdness. The following query still does not return all permutations

```scheme
  (Bind
    (Present (List (Variable "$Z") (Set (Variable "$X") (Variable "$Y"))))
    (Ordered (Variable "$Z") (Implication (Variable "$X") (Variable "$Y")))))
```

however, that one (`Ordered` has been replaced by `List`), does!!!

```scheme
  (Bind
    (Present (List (Variable "$Z") (Set (Variable "$X") (Variable "$Y"))))
    (List (Variable "$Z") (Implication (Variable "$X") (Variable "$Y")))))
```

**Comment by ngeiswei** [comment]
I've added the following (disabled and failing) utest to reproduce the bug

https://github.com/opencog/atomspace/blob/ad843e6e578dfbfc8955c3e2ce669dcbec564622/tests/query/PermutationsUTest.cxxtest#L82-L105

**Comment by linas** [comment]
Yikes!  Who should look into this?

**Comment by ngeiswei** [comment]
I'm looking into it right now, but if you've got some light-bulb idea I welcome it.

**Comment by linas** [comment]
Nope. Lots of prints. The permutation exploration code is pretty clean, but it is complex, and it does have some non-obvious termination conditions that are maybe off-by-one.  The difference between `List` and `Ordered` is very surprising... I'd start trying to figure that out first, before trying to figure out how permutations are explored.  Maybe fixing that is enough to get everything to work...

**Comment by ngeiswei** [comment]
I found the problem, not sure yet how to address it but here it is.

If `List` is use in the rewrite term instead of `Ordered`, the rarest link type in the clause becomes `Set`, used only twice, once in the clause, once in the KB.

If `Ordered` is used, then both `List` and `Set` are the rarest and `InitiateSearchCB::find_rarest` returns the whole clause as initial term. Since the top link in that term is ordered, the following section of code, trying permutations of unordered links, is never reached

https://github.com/opencog/atomspace/blob/ad843e6e578dfbfc8955c3e2ce669dcbec564622/opencog/query/PatternMatchEngine.cc#L1339-L1349

I suppose a workaround would be to force `find_rarest` to return the deepest unordered link, but that might compromise performances. The correct fix is to explore the permutations of deeper unordered links. I'm not sure yet where that should go, need to study the code more. Obviously any tip is welcome.

**Comment by linas** [comment]
OK. So, starting a search above a SetLink, and moving down into it, should have still explored all of the permutations. `tree_compare`  should have called `unordered_compare` which should have walked over the permutations.  I'll try for a few hours...

**Comment by linas** [comment]
It enters the permutation exploration code just fine, it knows its working on the first permutation, and then... its done!  It seems to have forgotten to explore the rest.

**Comment by linas** [comment]
Ahh! Easy Because it started with the top-level clause, it never explores more permutations.  I think that's because the code for the top-level loop is very old, before the unordered-link support was added, and it assumes that there can only be one grounding per top-level clause. Or something like that ... looking.

**Comment by linas** [comment]
Found it.  Need to alter the while-loop here:

https://github.com/opencog/atomspace/blob/ad843e6e578dfbfc8955c3e2ce669dcbec564622/opencog/query/PatternMatchEngine.cc#L2259-L2266

so that it's like this:
```
while (not found and (has_unordered and more_permutations...) OR (has_globs  ...))
```

**Comment by linas** [comment]
or not quite that. I mis-read my printouts... but its something close to that...

**Comment by linas** [comment]
Yes, that. Except that during the loop, the search never advances to the next permutation; it keeps retrying the first permutation.  I think I know why... not sure how to fix it...

**Comment by ngeiswei** [comment]
Awesome! Thanks, that's very helpful, should be able to take it from there.

**Comment by linas** [comment]
OK, so pull req #2348 fixes the reported bug.  However, there's more. This fails:
```
(List (Concept "A")
   (Set (Predicate "P") (Predicate "Q"))
   (Set (Predicate "R") (Predicate "S")))

(define q2
  (Bind
    (Present (List (Variable "$C")
       (Set (Variable "$X") (Variable "$Y"))
       (Set (Variable "$Z") (Variable "$W"))))
    (Implication
      (Implication (Variable "$X") (Variable "$Y"))
      (Implication (Variable "$Z") (Variable "$W")))))

(cog-execute! q2)
```
It finds only two permutations, instead of the expected four.  It's likely that nesting might also fail.

**Comment by linas** [comment]
@ngeiswei can you add unit tests for more complex cases?  e.g. three variables (should have 6 permutations) and nested sets (sets within sets...)  I'm going to spend a bit more time trying to fix the above; I believe that the  fix is to have the `_have_more` flag on a stack, along with the other permutation state...

**Comment by ngeiswei** [comment]
Sure, I'll do that on Monday though.

**Comment by linas** [comment]
Closing, this original bug report is fixed in #2357 although other issues remain: the current fix finds the wrong number of solutions for the pattern miner query https://github.com/opencog/atomspace/pull/2357#issuecomment-552384576

**Comment by linas** [comment]
@ngeiswei  FYI, I am fully and completely done with all know UnorderedLink pattern matcher issues.  I'm just writing here so that you see this post.  I'm not planning any more work; everything was fixed in #2388 

There are still two possible problem areas, but I do not want to look at them now; I've had my fill.

* Mixtures of Globs and Unordered links. I think some basic unit tests pass, but no one has pushed hard to explore complicated mixtures  (i.e. multiple unordered links, possibly nested, possibly as siblings, possibly with constraints with the same glob in multiple locations)

* Mixtures of unordered links and choice links. Choice links behave a lot like unordered links. I think some basic unit tests pass ... as far as I understand, neither the pattern miner, nor the URE use choice links, right?

**Comment by ngeiswei** [comment]
Thanks for the heads-up, and all the work involved in fixing that bug.

Mixtures of globs and unordered links will be heavily used once the unifier supports globs, we'll keep that in mind and complete the unit tests as necessary.

ChoiceLink is indeed not used by the pattern miner, the URE or PLN, maybe it should (potentially allowing to merge multiple rules into one, I suppose).
